pkg/base/helm.go

```diff
@@ -243,57 +254,168 @@ func removeCommonPrefix(baseFiles []BaseFile) []BaseFile {
 		}
+		upstreamToBasePathsMap[upstreamPath] = basePaths
+	}
+	return upstreamToBasePathsMap
+}
 
-		if base.Path == "" && upstreamFile.Path == path.Join(base.Path, "Chart.lock") {
-			base.AdditionalFiles = append(base.AdditionalFiles, BaseFile{
-				Path:    "Chart.lock",
-				Content: upstreamFile.Content,
-			})
-		} else if base.Path != "" && strings.Contains(upstreamFile.Path, path.Join(base.Path, "Chart.lock")) {
-			base.AdditionalFiles = append(base.AdditionalFiles, BaseFile{
-				Path:    "Chart.lock",
-				Content: upstreamFile.Content,
-			})
+// creates a map of the base paths and their cooresponding upstream paths
+func getBaseToUpstreamPathMap(upstreamFileMap map[string][]byte, upstreamToBasePathsMap map[string][]string) map[string]string {
+	baseToUpstreamPathMap := make(map[string]string)
+	for upstreamFilePath, basePaths := range upstreamToBasePathsMap {
+		for _, basePath := range basePaths {
+			baseToUpstreamPathMap[basePath] = upstreamFilePath
 		}
 	}
+	return baseToUpstreamPathMap
+}
+
+// adds the Chart.yaml and Chart.lock files from the upstream to the base charts that were rendered
+func helmChartBaseAppendAdditionalFiles(base Base, fullBasePath string, upstreamFiles map[string][]byte, baseToUpstreamPathMap map[string]string) Base {
+	if upstreamPath, ok := baseToUpstreamPathMap[fullBasePath]; ok {
+		additionalFiles := []string{"Chart.yaml", "Chart.lock"}
+		for _, additionalFile := range additionalFiles {
+			additionalFilePath := path.Join(upstreamPath, additionalFile)
+			if content, ok := upstreamFiles[additionalFilePath]; ok {
+				base.AdditionalFiles = append(base.AdditionalFiles, BaseFile{
+					Path:    additionalFile,
+					Content: content,
+				})
+			}
+		}
+	} else {
+		logger.Errorf("failed to find upstream path for base path %s", fullBasePath)
+	}
 
 	var nextBases []Base
-	for _, base := range base.Bases {
-		base = helmChartBaseAppendAdditionalFiles(base, u)
-		nextBases = append(nextBases, base)
+	for _, nextBase := range base.Bases {
+		nextBase = helmChartBaseAppendAdditionalFiles(nextBase, path.Join(base.Path, nextBase.Path), upstreamFiles, baseToUpstreamPathMap)
+		nextBases = append(nextBases, nextBase)
 	}
 	base.Bases = nextBases
 
 	return base
 }
 
-// look for any sub-chart dependencies that are missing from base and add their Chart.yaml
-func helmChartBaseAppendMissingDependencies(base Base, u *upstreamtypes.Upstream) Base {
-	basePaths := getAllBasePaths("", base)
-	basePathMap := map[string]bool{}
-	for _, basePath := range basePaths {
-		basePathMap[basePath] = true
+func helmChartUpstreamPathToBasePaths(upstreamPath string, upstreamFileMap map[string][]byte) ([]string, error) {
+	charts := pathToCharts(upstreamPath)
+	basePathParts := make([][]string, len(charts))
+	basePathParts[0] = []string{""}
+	// iterate over the subcharts and find any that are aliased
+	for i := 1; i < len(charts); i++ {
+		parts := []string{}
+		chart := charts[i]
+		parts = append(parts, chart)
```

Comment:
this would break if i only have aliases for this particular chart. i think this should be done [here](https://github.com/replicatedhq/kots/pull/3025/files?diff=split&w=0#diff-cc6bf3b28c8d864ab3c297aa234a72fc8a7e035ae541171f91de94971a10c40aR336-R338) if the chart exists but doesn't have an alias

Reply:
If there is an upstream path for a chart/subchart, then it's dependency, but it is not aliased (or was aliased manually by the name of the subdirectory).  Dependencies can be managed either by including them in the `charts/` directory or in the `dependencies` field (which *is* where they might be aliased).  This is why I add the actual chart name here by default and then add any aliases for it later.

This is what I gathered from reading this: https://helm.sh/docs/topics/charts/#chart-dependencies

Reply:
Addressed this by using the dependencies list from the parent chart to determine the aliases (if any).  If the subchart is not present in the parent chart's dependencies list, then the chart name is used.